### ISSUE TITLE
Dataclass InitVar is reported as unknown

### ISSUE THREAD

**Issue by vberlier** [microsoft/pyright#4091]
I'm using Pyright `v1.1.276`. In the following snippet the `InitVar` triggers an error:

```py
from dataclasses import InitVar, dataclass


@dataclass
class Thing:
    foo: InitVar[int] = 123
    # error: Type of "foo" is unknown (reportUnknownVariableType)

    def __post_init__(self, foo: int):
        print(foo)
```

**Comment by erictraut** [comment]
I presume that you have enabled "strict" mode because the `reportUnknownVariableType` rule is not enabled by default in pyright.

If I enable strict-mode checks, I'm still not able to repro the problem you're reporting here. Can you verify that the code snippet above results in an error? Are there any other steps I need to take to repro?

One possibility is that you have an old `dataclasses` library installed in your Python environment that is overriding the stdlib version. If that's the case, please uninstall the third-party library.

**Comment by vberlier** [comment]
Yeah I'm using strict mode. I don't have any special `dataclasses` library installed but I'm on python 3.8, that might be the issue.

**Comment by erictraut** [comment]
Dataclass was introduced in python 3.7, so using 3.8 shouldn't be an issue. I'm pretty confident that this isn't a bug in pyright, but I'm still not able to repro the problem, so I'm not sure what's different about your configuration. 

Do you have any custom typeshed stubs installed?

Are you able to repro if you create a fresh virtual environment?

If you right click on `InitVar` and choose "Go To Declaration", where does it take you? I would expect it to go to `dataclasses.pyi`, a file that is bundled with pyright.

**Comment by vberlier** [comment]
When I go to declaration I end up at the `InitVar` class in `/usr/lib/python3.8/dataclasses.py`

```py
class _InitVarMeta(type):
    def __getitem__(self, params):
        return InitVar(params)

class InitVar(metaclass=_InitVarMeta):
    __slots__ = ('type', )

    def __init__(self, type):
        self.type = type

    def __repr__(self):
        if isinstance(self.type, type):
            type_name = self.type.__name__
        else:
            # typing objects, e.g. List[int]
            type_name = repr(self.type)
        return f'dataclasses.InitVar[{type_name}]'
```

I can repro in a fresh virtualenv

```bash
$ cd /tmp
$ mkdir deleteme
$ cd deleteme
$ python -m venv .venv
$ source .venv/bin/activate
$ cat > pyproject.toml
[tool.pyright]
typeCheckingMode = "strict"

$ cat > demo.py
from dataclasses import InitVar, dataclass


@dataclass
class Thing:
    foo: InitVar[int] = 123
    # error: Type of "foo" is unknown (reportUnknownVariableType)

    def __post_init__(self, foo: int):
        print(foo)

$ npm i -D pyright@latest
$ npx pyright
No configuration file found.
pyproject.toml file found at /tmp/deleteme.
Loading pyproject.toml file at /tmp/deleteme/pyproject.toml
Assuming Python version 3.8
Assuming Python platform Linux
No include entries specified; assuming /tmp/deleteme
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
stubPath /tmp/deleteme/typings is not a valid directory.
Searching for source files
Found 1 source file
pyright 1.1.276
/tmp/deleteme/demo.py
  /tmp/deleteme/demo.py:6:5 - error: Type of "foo" is unknown (reportUnknownVariableType)
1 error, 0 warnings, 0 informations
Completed in 0.604sec
```

**Comment by erictraut** [comment]
OK, thanks for the additional info. I'm able to repro the problem, and I've found the change the caused it. It's due to [a recent change](https://github.com/python/typeshed/pull/8852) in the typeshed stub file `dataclasses.pyi`. Someone added a metaclass for the `InitVar` class:

```python
if sys.version_info >= (3, 9):
    _InitVarMeta: TypeAlias = type
else:
    class _InitVarMeta(type):
        # Not used, instead `InitVar.__class_getitem__` is called.
        def __getitem__(self, params: Any) -> InitVar[Any]: ...

class InitVar(Generic[_T], metaclass=_InitVarMeta):
    ...
```

The problem here is the introduction of `_InitVarMeta.__getitem__` for Python 3.8. This is confusing pyright's logic for `InitVar`. I'll need to dig deeper to find a solution.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.277, which I just published. It will also be included in a future release of pylance.
